docs/design-tech/sync.md

```diff
@@ -1,89 +1,263 @@
-# Syncing within core
+# Sync v3
+## Overview
+Sync v3 is the latest iteration of our logic that syncs users' file trees across devices. It is designed under the following key constraints:
+* The state of a user's file tree must satisfy certain invariants at all times, on all devices (including the server), even in the event of process interruptions and failures.
+* Users' file contents and names must not be readable by the server, but users can share files and folders with other users.
+* Users must be able to view and edit files offline, then sync frictionlessly, even with intervening edits made to files on other devices.
+* The performance of the system must not suffer as a result of typical usage, which presents a number of challenges worth discussion
 
-Understanding the [data model](data_model.md) is a prerequisite to understanding this document.
+These design constraints motivate the following high-level design, which will be flushed out in more detail after a further discussion of the constraints:
+* The client maintains two versions of the file tree: one for the current local version (`local`), and one for the last version known to be agreed upon by the client and server (`base`).
+* The server maintains one version of each user's file tree (`remote`)
+* To perform a sync, the client pulls `remote` from the server, 3-way merges the changes with `local` and `base`, and pushes those updates to the server.
 
-Sync is the means by which a user updates a client with either their changes on other devices or peer changes on shared files.
+## Constraints
+### File Tree Invariants
+Before discussing the invariants, it is helpful to understand the data model. For the purposes of Sync v3, each file has:
+* `id`: an immutable unique identifier
+* `file_type`: an immutable flag that indicates if the file is a folder or document. A _file_ is either a _document_, which has contents, or a _folder_, which can have files as children.
+* `name`: the file's name. Clients encrypt and HMAC names before sending them to the server, so while the server cannot read file names, it can test them for equality.
+* `parent`: the file's parent folder (specified by the parent's `id`). The `root` folder is the folder which has itself as a parent (it's name is by convention it's owners username and it cannot be deleted).
+* `deleted`: a boolean flag which indicates whether the file is _explicitly_ deleted. A file whose ancestor is deleted need not be explictly deleted; if not, it is considered _implictly_ deleted. A file that is no longer known to a client at all is considered _pruned_ on that client.
```

Comment:
`need not be explictly deleted` doesn't the server have too delete all children explicitly tho?

Reply:
yeah in the meantime it is not explicitly deleted